File: services/typescript/fastify/src/routes/handlers/get.ts
import type { RouteHandler } from "fastify";
import { z } from "zod";
import type { State } from "../../state.js";
import { getLogger } from "../../logger.js";
import { AppErrors, sendError } from "../errors.js";
import type { GetRoute } from "../types.js";

export function getHandler(state: State): RouteHandler<GetRoute> {
  return async (request, reply) => {
    const logger = getLogger();

    const id = z.uuidv4().safeParse(request.params.id);
    if (!id.success) {
      logger.warn(
        { method: request.method, path: request.url },
        "Invalid path"
      );
      return sendError(reply, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);
    if (!task) {
      logger.warn(
        { id: id.data, method: "GET", path: request.url },
        "Task not found"
      );
      return sendError(reply, AppErrors.TaskNotFound);
    }

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "GET"
      },
      "Retrieved task"
    );

    return reply.status(200).send(task);
  };
}

File: services/python/fastapi/src/routes/handlers/get.py
import structlog
from uuid import UUID
from fastapi import Request
from fastapi.responses import JSONResponse

from app.state import AppState
from app.task import Task
from routes.errors import send_error, AppErrors


async def get_handler(request: Request, id: UUID) -> JSONResponse | Task:
    logger = structlog.get_logger()

    state: AppState = request.app.state.app_state

    async with request.app.state.tasks_lock:
        task: Task | None = state.tasks.get(id)

        if task is None:
            logger.warn(
                "Task not found", id=str(id), method="GET", path=request.url.path
            )
            return send_error(AppErrors.TaskNotFound)

    logger.info(
        "Retrieved task",
        id=str(id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method="GET",
    )

    return task
